core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolverGroup.java

```diff
@@ -127,6 +133,9 @@
         assert executor instanceof EventLoop;
         final EventLoop eventLoop = (EventLoop) executor;
         final DnsNameResolverBuilder builder = new DnsNameResolverBuilder(eventLoop);
+        builder.dnsQueryLifecycleObserverFactory(
```

Comment:
I think we should set this in `DnsResolverGroupBuilder` so that this factory is not overridden by the factory that the user configures.

Reply:
Can you please elaborate on this? You mean set the dnsQueryLifecycleObserverFactory in the DnsResolverGroupBuilder::build method ?

Reply:
Sure. 😄 
After we set the observer factory `resolverConfigurator.accept(builder);` at line 139 is called.
In that method, another observer factory might be set 
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java#L332
So I think we can do sort of:
```java
final DnsQueryLifecycleObserverFactory defaultObserverFactory =
        new DefaultDnsQueryLifecycleObserverFactory(
                meterRegistry, new MeterIdPrefix("armeria.client.dns.queries")));
if (dnsQueryLifecycleObserverFactory == null) {
    builder.dnsQueryLifecycleObserverFactory(defaultObserverFactory);
} else {
    builder.dnsQueryLifecycleObserverFactory(
            new SomeWrapper(defaultObserverFactory, dnsQueryLifecycleObserverFactory));
}
```

Reply:
And we might need to add a boolean parameter to the setter so that a user can choose the behavior?
`DnsResolverGrouBuilder.dnsQueryLifecycleObserverFactory(observerFactory, overrideDefaultFactory)`
I'm not sure if it's a good idea though. 😅 
Do you have any suggestion? @trustin, @ikhoon

Reply:
@trustin @ikhoon  can you please comment when you get a chance?

Reply:
Sorry for the late reply. Let me take a look.

Reply:
How about moving the modifications in this class to `DnsResolverGroupBuilder`, so that:

- `RefreshingAddressResolverGroup` doesn't change.
- `DnsResolverGroupBuilder` creates the default observer factory;
- `DnsResolverGroupBuilder.build()` creates the `resolverConfigurator` that calls `DnsNameResolverBuilder.dnsQueryLifecycleObserverFactory()` with the default observer factory.
  - Use `io.netty.resolver.dns.BiDnsQueryLifecycleObserverFactory` to combine the default observer factory and the user-specified observer factory.

Makes sense?

Reply:
I didn't know there was `BiDnsQueryLifecycleObserverFactory` 😄